plugins/database/mongodb/connection_producer.go

```diff
@@ -106,120 +135,49 @@ func (c *mongoDBConnectionProducer) Init(ctx context.Context, conf map[string]in
 
 // Connection creates or returns an existing a database connection. If the session fails
 // on a ping check, the session will be closed and then re-created.
-func (c *mongoDBConnectionProducer) Connection(_ context.Context) (interface{}, error) {
+func (c *mongoDBConnectionProducer) Connection(ctx context.Context) (interface{}, error) {
 	if !c.Initialized {
 		return nil, connutil.ErrNotInitialized
 	}
 
-	if c.session != nil {
-		if err := c.session.Ping(); err == nil {
-			return c.session, nil
+	if c.client != nil {
+		if err := c.client.Ping(ctx, readpref.Primary()); err == nil {
+			return c.client, nil
 		}
-		c.session.Close()
+		// Ignore error on purpose since we want to re-create a session
```

Comment:
This code seems a little weird to me. The `mongo.Client` type is resilient to failovers, so a failed `Ping` does not necessarily indicate that the client should be restarted. If the server is in the cluster of failing over during the `Ping`, the `Client` will re-calibrate its view of the cluster and figure out what changed.

Reply:
> If the server is in the cluster of failing over during the Ping, the Client will re-calibrate its view of the cluster and figure out what changed.

I assume this happens automatically while we ping the server, correct? If that is the case I think this is just for safety reasons in case something bad happens. I agree that mostly the next connection attempt will fail too but I also think it doesn't hurt to try it again?

Reply:
Instead of creating a completely new `Client`, maybe we can just try the `Ping` again? This is the biggest downfall of using the `Ping` method in general: it's good for testing server connectivity but if the server is undergoing a transient state change, it will report a failure even though it's not really a failure.

If you want to avoid risk, though, perhaps it is best to keep the existing behavior and just recreate the `Client`. I'll defer that to you.

Reply:
I personally think that creating a fresh client which adds a minimal overhead compared to reusing the old client is the better choice here.